_posts/2020-10-19-tests-best-practices.md

```diff
@@ -0,0 +1,233 @@
+fixtures and not in the helper libraries it calls.
+
+Embedding assertions in helper libraries does not allow tests to control
+how to react to errors or specific responses, limiting the usage of such
+helpers by tests and the information that can be provided (data may have
+been collected on the way back of the stack which could be relevant for
+debugging).
+
+Note: Custom assertion wrappers are considered assertions. These are focused
+on assertion and named accordingly.
+
+### Traceability on Failure
+Upon assert failure, leave enough information that explains what failed.
+This is commonly supported by the test name, the asserted objects and their
+content.
+
+The reported failure should point back to the source code (e.g. line number,
+traceback) to ease debugging.
+
+### Shared Resources
+Shared resources between tests have the potential to leak state changes
+between tests, violating the test isolation.
+
+Therefore, whenever possible it would be always preferable to recreate any
+resource from scratch for each and every test.
+
+Said that, one cannot ignore that some tests (e.g. e2e) pay a heavy cost
+when working with resources (reflected by time and/or memory).
+
+In such scenarios, a more pragmatic approach is in order.
+Sharing resources across a group of tests with an immutable restriction many
+times worth the management effort and results with faster and less costly
+tests.
+
+### Continue On Failure
+Expect tests to fail, stopping on the first failure may not be informative
+enough.
+Said that, leave behind enough information to debug and troubleshoot the
+failure.
+
+### Avoid Dead Tests
+Production code has `*tests`. Tests have only themselves.
+Do not leave unexecuted code, it rots.
+Keep only running tests and helper functions.
+
+Note: Tests that are tracking failing scenarios (bugs, missing features) are
+better executed with an expected failure over skipped.
+
+
+## Parallel Tests
+
+Running tests in parallel can decrease substantially the tests runtime if
+the tests are mostly I/O bound.
+
+When tests are running in parallel against a single SUT, several points are
+in practice added to the tests:
+- Checking how well a test is isolated, as it may now randomly be executed with
+  other tests in parallel. The mode in which the tests are executed in parallel
+  affect the risk of this point (e.g. multi threaded or multi process).
+- Checking how well the SUT is behaving while multiple clients (i.e. the tests)
+  challenge its API in parallel.
+- In both points above, the randomization is an important factor.
+- System logs may be now difficult to examine when debugging and
+  troubleshooting failures.
+
+Therefore, an amount of uncertainty is introduced, something that tests try to
+avoid in order to be reproducible and consistent in their reported results.
+
+Projects should balance between efficient runtime, covering concurrent access
+to the SUT and the risk of uncertainty and debugging effort.
+
+
+## Workarounds
```

Comment:
in this section you mention parallel tests, which make me wonder if this section should be s sub-section of `## Parallel Tests`? if it is, then maybe change the the title to `### Workarounds`.